src/crimson/os/seastore/lba_mapping.h

```diff
@@ -10,125 +10,198 @@
-    return pos;
+  static LBAMapping create_indirect(LBACursorRef iter) {
+    return LBAMapping(nullptr, std::move(iter));
   }
 
-  extent_len_t get_length() const {
-    ceph_assert(range.end > range.begin);
-    return len;
+  LBAMapping(const LBAMapping &) = delete;
+  LBAMapping(LBAMapping &&) = default;
+
+  LBAMapping &operator=(const LBAMapping &) = delete;
+  LBAMapping &operator=(LBAMapping &&) = default;
+
+  ~LBAMapping() = default;
+
+  bool is_null() const {
+    return !physical_cursor && !indirect_cursor;
   }
 
-  paddr_t get_val() const {
-    return value.get_paddr();
+  bool is_indirect() const {
+    assert(!is_null());
+    return (bool)indirect_cursor;
   }
 
-  virtual laddr_t get_key() const {
-    return range.begin;
+  bool is_valid() const {
+    assert(!is_null());
+    return physical_cursor->is_valid()
+	&& (!is_indirect() || indirect_cursor->is_valid());
   }
 
-  bool has_been_invalidated() const {
-    return parent->has_been_invalidated();
+  // For reserved mappings, the return values are
+  // undefined although it won't crash
+  bool is_stable() const;
+  bool is_data_stable() const;
+  bool is_clone() const {
+    assert(!is_null());
+    return physical_cursor->val.refcount > 1;
+  }
+  bool is_zero_reserved() const {
+    assert(!is_null());
+    return !get_val().is_real();
   }
 
-  bool is_parent_viewable() const {
-    ceph_assert(parent);
-    auto [viewable, state] = parent->is_viewable_by_trans(
-      ctx.trans.get_trans_id());
-    if (state == CachedExtent::viewable_state_t::stable_retired) {
-      assert(ctx.trans.is_retired(parent->get_paddr(), parent->get_length()));
+  extent_len_t get_length() const {
+    assert(!is_null());
+    if (is_indirect()) {
+      return indirect_cursor->val.len;
+    } else {
+      return physical_cursor->val.len;
     }
-    return viewable;
   }
-  bool is_parent_valid() const {
-    ceph_assert(parent);
-    return parent->is_valid();
+
+  paddr_t get_val() const {
+    assert(!is_null());
+    return physical_cursor->val.pladdr.get_paddr();
+  }
+
+  uint32_t get_checksum() const {
+    assert(!is_null());
+    return physical_cursor->val.checksum;
   }
 
-  virtual void maybe_fix_pos() = 0;
-  virtual bool parent_modified() const = 0;
-  virtual uint32_t get_checksum() const = 0;
+  laddr_t get_key() const {
+    assert(!is_null());
+    if (is_indirect()) {
+      return indirect_cursor->key;
+    } else {
+      return physical_cursor->key;
+    }
+  }
 
   // An lba pin may be indirect, see comments in lba_manager/btree/btree_lba_manager.h
-  virtual bool is_indirect() const = 0;
-  virtual laddr_t get_intermediate_key() const = 0;
-  virtual laddr_t get_intermediate_base() const = 0;
-  virtual extent_len_t get_intermediate_length() const = 0;
-  // The start offset of the pin, must be 0 if the pin is not indirect
-  virtual extent_len_t get_intermediate_offset() const = 0;
+  laddr_t get_intermediate_key() const {
+    assert(!is_null());
```

Comment:
Seems to me **most of** the `assert(!is_null())` should be replaced by the more accurate `assert(is_linked_direct())`.